linera-service/src/util.rs

```diff
@@ -1,132 +1,80 @@
 // Copyright (c) Zefchain Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use anyhow::{bail, Result};
-use once_cell::sync::OnceCell;
-use std::{
-    collections::{HashMap, HashSet},
-    env,
-    path::{Path, PathBuf},
-};
-use tokio::{process::Command, sync::Mutex};
-use tracing::info;
+use anyhow::{anyhow, bail, Result};
+use std::path::{Component, PathBuf};
+use tokio::process::Command;
+use tracing::{debug, error};
 
-/// The name of the environment variable that allows specifying additional arguments to be passed
-/// to `cargo` when starting client, server and proxy processes.
-const CARGO_ENV: &str = "LINERA_CARGO_PARAMS";
+/// Attempts to resolve the path and test the version of the given binary against our
+/// package version. This is meant for binaries of the Linera repository.
+pub async fn resolve_binary(name: &'static str, package: &'static str) -> Result<PathBuf> {
+    debug!("Resolving binary {name} based on the current binary path.");
+    let mut current_binary_path = PathBuf::from(
+        std::env::args()
+            .next()
+            .expect("args should start with the current binary path"),
+    );
+    current_binary_path.pop();
 
-pub fn resolve_cargo_binary(bin_name: &'static str) -> Result<PathBuf> {
-    cargo_home().map(|cargo_home| cargo_home.join("bin").join(bin_name))
-}
-
-fn cargo_home() -> Result<PathBuf> {
-    if let Ok(cargo_home) = std::env::var("CARGO_HOME") {
-        Ok(PathBuf::from(cargo_home))
-    } else if let Some(home) = dirs::home_dir() {
-        Ok(home.join(".cargo"))
-    } else {
-        bail!("could not find CARGO_HOME directory, please specify it explicitly")
-    }
-}
-
-#[allow(unused_mut)]
-fn detect_current_features() -> Vec<&'static str> {
-    let mut features = vec![];
-    #[cfg(feature = "benchmark")]
-    {
-        features.push("benchmark");
-    }
-    #[cfg(feature = "wasmer")]
-    {
-        features.push("wasmer");
-    }
-    #[cfg(feature = "wasmtime")]
-    {
-        features.push("wasmtime");
-    }
-    #[cfg(feature = "rocksdb")]
-    {
-        features.push("rocksdb");
-    }
-    #[cfg(feature = "scylladb")]
+    #[cfg(any(test, feature = "test"))]
+    // Test binaries are typically in target/debug/deps while crate binaries are in target/debug
+    // (same thing for target/release).
     {
-        features.push("scylladb");
+        let components = current_binary_path
```

Comment:
This is a little harder to read, consider: https://doc.rust-lang.org/std/path/struct.Path.html#method.ends_with

```rust
 if path.ends_with(Path::new("target/debug/deps")) || path.ends_with(Path::new("target/release/deps")) {
```

Reply:
nice